drv/fpga-devices/src/ecp5.rs

```diff
@@ -0,0 +1,496 @@
+    ReadUserCode = 0xc0,
+    ReadStatus = 0x3c,
+    CheckBusy = 0xf0,
+    Refresh = 0x79,
+    EnableConfigurationMode = 0xc6,
+    EnableTransparentConfigurationMode = 0x74,
+    DisableConfigurationMode = 0x26,
+    Erase = 0x0e,
+    BitstreamBurst = 0x7a,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Trace {
+    None,
+    Disabled,
+    Enabled,
+    Command(Command),
+    ReadId(Id),
+    Read32(Command, u32),
+    StandardBitstreamDetected,
+    EncryptedBitstreamDetected,
+    WriteBitstreamChunk,
+    BitstreamError(BitstreamError),
+    UserDesignResetAsserted,
+    UserDesignResetDeasserted,
+    Lock,
+    Unlock,
+}
+ringbuf!(Trace, 16, Trace::None);
+
+pub const DEVICE_RESET_DURATION: u64 = 25;
+pub const USER_DESIGN_RESET_DURATION: u64 = 25;
+pub const BUSY_DURATION: u64 = 10;
+pub const DONE_DURATION: u64 = 10;
+
+pub trait Ecp5Driver {
+    type Error: Debug;
+
+    /// PROGAM_N interface. This pin acts as a device reset and when asserted
+    /// low force to (re)start the bitstream loading process.
+    ///
+    /// See FPGA-TN-02039-2.0, 4.5.2 for details.
+    fn program_n(&self) -> Result<bool, Self::Error>;
+    fn set_program_n(&self, asserted: bool) -> Result<(), Self::Error>;
+
+    /// INIT_N interface. This pin can be driven after reset/power up to keep
+    /// the device from entering Configuration state. As input it signals
+    /// Initialization complete or an error occured during bitstream loading.
+    ///
+    /// See FPGA-TN-02039-2.0, 4.5.3 for details.
+    fn init_n(&self) -> Result<bool, Self::Error>;
+    fn set_init_n(&self, asserted: bool) -> Result<(), Self::Error>;
+
+    /// DONE interface. This pin signals the device is in User Mode. Asserting
+    /// the pin keeps the device from entering User Mode after Configuration.
+    ///
+    /// See FPGA-TN-02039-2.0, 4.5.4 for details.
+    fn done(&self) -> Result<bool, Self::Error>;
+    fn set_done(&self, asserted: bool) -> Result<(), Self::Error>;
+
+    /// A generic interface to send commands and read/write data from a
+    /// configuration port. This interface is intended to be somewhat transport
+    /// agnostic so either SPI or JTAG could be implemented if desired.
+    fn configuration_read(&self, data: &mut [u8]) -> Result<(), Self::Error>;
+    fn configuration_write(&self, data: &[u8]) -> Result<(), Self::Error>;
+    fn configuration_write_command(
+        &self,
+        c: Command,
+    ) -> Result<(), Self::Error>;
+
+    /// The configuration interface may exist on a shared medium such as SPI.
+    /// The following primitives allow the upper half of the driver to issue
+    /// atomic commands.
+    ///
+    /// If no lock control of the medium is needed these can be implemented as
+    /// no-op.
+    fn configuration_lock(&self) -> Result<(), Self::Error>;
+    fn configuration_release(&self) -> Result<(), Self::Error>;
+
+    /// User design reset.
+    fn user_design_reset_n(&self) -> Result<bool, Self::Error>;
```

Comment:
There's a lot of overlap between this trait and `FpgaUserDesign` below.

Could we instead define the trait as
```rust
pub trait Ecp5Driver : FpgaUserDesign {
    ...
```
and remove some duplication (since that would force anyone implementing `Ecp5Driver` to also implement `FpgaUserDesign`)?